compiler/AST/AstToText.cpp

```diff
@@ -0,0 +1,1318 @@
+  }
+  else
+  {
+    INT_ASSERT(false);
+  }
+}
+
+/************************************ | *************************************
+*                                                                           *
+* Append a normalized version of the formals list to the buffer.            *
+*                                                                           *
+************************************* | ************************************/
+
+void AstToText::appendFormals(FnSymbol* fn)
+{
+  int  count = numFormals(fn);
+  bool skip  = skipParens(fn);
+  bool first = true;
+
+  if (skip == false)
+    mText += '(';
+
+  for (int index = 1; index <= count; index++)
+  {
+    ArgSymbol* arg = formalGet(fn, index);
+
+    if (arg->hasFlag(FLAG_IS_MEME) == false)
+    {
+      if (first == true)
+      {
+        if (skip == true)
+          mText += " ";
+
+        first = false;
+      }
+      else
+        mText += ", ";
+
+      appendFormal(fn, index);
+    }
+  }
+
+  if (skip == false)
+    mText += ')';
+}
+
+bool AstToText::skipParens(FnSymbol* fn) const
+{
+  bool retval = false;
+
+  if (fn->hasFlag(FLAG_NO_PARENS))
+    retval = true;
+
+  else if (fn->hasFlag(FLAG_TYPE_CONSTRUCTOR) && fn->numFormals() == 0)
+    retval = true;
+
+  else if (fn->hasFlag(FLAG_MODULE_INIT)      && developer        == false)
+    retval = true;
+
+  else
+    retval = false;
+
+  return retval;
+}
+
+/************************************ | *************************************
+*                                                                           *
+* Append a normalized version of a formal to the buffer.                    *
+*                                                                           *
+************************************* | ************************************/
+
+// Excludes "_mt" (method token) and "this" when they are present
+int AstToText::numFormals(FnSymbol* fn) const
+{
+  int retval = fn->formals.length;
+
+  if (fn->isMethod() == true)
+    retval = retval - indexForThis(fn);
+
+  return retval;
+}
+
+void AstToText::appendFormal(FnSymbol* fn, int oneBasedIndex)
+{
+  ArgSymbol* arg = formalGet(fn, oneBasedIndex);
+
+  appendFormalIntent(arg);
+
+  appendFormalName(arg);
+
+  appendFormalType(arg);
+
+  appendFormalVariableExpr(arg);
+
+  appendFormalDefault(arg);
+}
+
+void AstToText::appendFormalIntent(ArgSymbol* arg)
+{
+  switch (arg->intent)
+  {
+    case INTENT_IN:
+      mText += "in ";
+      break;
+
+    case INTENT_OUT:
+      mText += "out ";
+      break;
+
+    case INTENT_INOUT:
+      mText += "inout ";
+      break;
+
+    case INTENT_CONST:
+      mText += "const ";
+      break;
+
+    case INTENT_CONST_IN:
+      mText += "const in ";
+      break;
+
+    case INTENT_REF:
+      mText += "ref ";
+      break;
+
+    case INTENT_CONST_REF:
+      mText += "const ref ";
+      break;
+
+    case INTENT_PARAM:
+      mText += "param ";
+      break;
+
+    case INTENT_TYPE:
+      break;
```

Comment:
Is there a case where the arg would have INTENT_TYPE but not have FLAG_TYPE_VARIABLE?  Should we check to be certain?